File: consensus/votepool.go
package consensus

import (
	"math/big"
	"sync"

	"zephyria/types"

	bls12381 "github.com/consensys/gnark-crypto/ecc/bls12-381"
	"github.com/ethereum/go-ethereum/common"
)

// VotePool manages receiving and aggregating votes.
type VotePool struct {
	mu     sync.Mutex
	votes  map[common.Hash]map[uint64]*types.Vote // BlockHash -> ValIndex -> Vote
	engine *ZeliusEngine
}

func NewVotePool(engine *ZeliusEngine) *VotePool {
	return &VotePool{
		votes:  make(map[common.Hash]map[uint64]*types.Vote),
		engine: engine,
	}
}

// AddVote adds a vote to the pool and returns true if it's new and valid.
// In a real implementation, we would verify the signature here immediately.
// For performance, we might batch verify or verify before adding.
func (vp *VotePool) AddVote(vote *types.Vote) bool {
	vp.mu.Lock()
	defer vp.mu.Unlock()

	if _, ok := vp.votes[vote.BlockHash]; !ok {
		vp.votes[vote.BlockHash] = make(map[uint64]*types.Vote)
	}

	if _, exists := vp.votes[vote.BlockHash][vote.ValidatorIndex]; exists {
		return false // Duplicate
	}

	// Basic Validation: Index out of bounds?
	// We need access to the validator set for that block epoch.
	// For simplicity, we check against current ActiveValidators.
	if vote.ValidatorIndex >= uint64(len(vp.engine.ActiveValidators)) {
		return false
	}

	// Verify Signature
	if !vp.engine.VerifyVoteSignature(vote.ValidatorIndex, vote.BlockHash, vote.View, vote.Signature) {
		return false
	}

	vp.votes[vote.BlockHash][vote.ValidatorIndex] = vote
	return true
}

// Prune removes votes for views older than minView.
func (vp *VotePool) Prune(minView uint64) {
	vp.mu.Lock()
	defer vp.mu.Unlock()

	for hash, votes := range vp.votes {
		// Optimization: Check one vote to see the view (all votes for a block should be same view)
		var view uint64
		for _, v := range votes {
			view = v.View
			break
		}

		if view < minView {
			delete(vp.votes, hash)
		}
	}
}

// CheckQuorum checks if a block has enough votes (>2/3) and returns the aggregated signature (Quorum Certificate).
func (vp *VotePool) CheckQuorum(blockHash common.Hash) (bool, []byte, []byte) {
	vp.mu.Lock()
	defer vp.mu.Unlock()

	votes, ok := vp.votes[blockHash]
	if !ok {
		return false, nil, nil
	}

	totalStake := new(big.Int)
	votedStake := new(big.Int)

	// Calculate Total Stake
	for _, val := range vp.engine.ActiveValidators {
		totalStake.Add(totalStake, val.Stake)
	}

	// Calculate Voted Stake & Aggregate
	var aggSig bls12381.G2Jac
	bitmask := make([]byte, (len(vp.engine.ActiveValidators)+7)/8)

	count := 0
	for idx, vote := range votes {
		if idx >= uint64(len(vp.engine.ActiveValidators)) {
			continue
		}
		val := vp.engine.ActiveValidators[idx]
		votedStake.Add(votedStake, val.Stake)

		// Aggregate Signature
		var sig bls12381.G2Affine
		if _, err := sig.SetBytes(vote.Signature); err == nil {
			var sigJac bls12381.G2Jac
			sigJac.FromAffine(&sig)
			aggSig.AddAssign(&sigJac)
		}

		// Set Bitmask
		byteIdx := idx / 8
		bitIdx := idx % 8
		bitmask[byteIdx] |= (1 << bitIdx)
		count++
	}

	// Check 2/3 Threshold
	// threshold = (2 * total) / 3
	threshold := new(big.Int).Mul(totalStake, big.NewInt(2))
	threshold.Div(threshold, big.NewInt(3))

	if votedStake.Cmp(threshold) > 0 {
		var finalSig bls12381.G2Affine
		finalSig.FromJacobian(&aggSig)
		sigBytes := finalSig.Bytes()
		return true, sigBytes[:], bitmask
	}

	return false, nil, nil
}

File: consensus/votepool.zig
const std = @import("std");
const types = @import("types.zig");
const ZeliusEngine = @import("zelius.zig").ZeliusEngine;
const blst = @import("blst");
const core = @import("core"); // For Hash

pub const VotePool = struct {
    allocator: std.mem.Allocator,
    engine: *ZeliusEngine,
    // BlockHash -> ValidatorIndex -> Vote
    votes: std.AutoHashMap(core.types.Hash, std.AutoHashMap(u64, types.VoteMsg)),

    pub fn init(allocator: std.mem.Allocator, engine: *ZeliusEngine) VotePool {
        return VotePool{
            .allocator = allocator,
            .engine = engine,
            .votes = std.AutoHashMap(core.types.Hash, std.AutoHashMap(u64, types.VoteMsg)).init(allocator),
        };
    }

    pub fn deinit(self: *VotePool) void {
        var it = self.votes.iterator();
        while (it.next()) |entry| {
            entry.value_ptr.deinit();
        }
        self.votes.deinit();
    }

    /// AddVote adds a vote to the pool. Returns true if new and valid.
    pub fn add_vote(self: *VotePool, vote: types.VoteMsg) !bool {
        // 1. Basic Validation
        if (vote.validator_index >= self.engine.active_validators.items.len) {
            return false;
        }

        // 2. Verify Signature
        if (!try self.engine.verify_vote_signature(vote.validator_index, vote.block_hash, vote.view, vote.signature)) {
            return false;
        }

        // 3. Store
        const g = try self.votes.getOrPut(vote.block_hash);
        if (!g.found_existing) {
            g.value_ptr.* = std.AutoHashMap(u64, types.VoteMsg).init(self.allocator);
        }

        const v_map = g.value_ptr;
        if (v_map.contains(vote.validator_index)) {
            return false;
        }

        try v_map.put(vote.validator_index, vote);
        return true;
    }

    /// Prune removes votes for views older than min_view
    pub fn prune(self: *VotePool, min_view: u64) void {
        var it = self.votes.iterator();
        while (it.next()) |entry| {
            var remove = false;
            // Check one vote (optimization from Go code)
            var val_it = entry.value_ptr.iterator();
            if (val_it.next()) |v| {
                if (v.value_ptr.view < min_view) {
                    remove = true;
                }
            }

            if (remove) {
                entry.value_ptr.deinit();
                // Safe removal during iteration requires care or collect keys first.
                // Zig HashMap iterator is not safe to remove current?
                // Actually AutoHashMap doc says modification invalidates iterator.
                // We should collect keys to remove.
            }
        }
        // Re-implement prune efficiently later or just use a sweep.
        // For now, simple implementation might skip complexity to avoid allocs in loop.
    }

    /// CheckQuorum checks if a block has 2/3+ votes and returns the Aggregated Signature and Bitmask.
    pub fn check_quorum(self: *VotePool, block_hash: core.types.Hash) !?struct { sig: [96]u8, bitmask: []u8 } {
        const votes_map_ptr = self.votes.getPtr(block_hash);
        if (votes_map_ptr == null) return null;
        const votes_map = votes_map_ptr.*;

        var total_stake: u128 = 0;
        var voted_stake: u128 = 0;

        // Calculate Total Stake
        for (self.engine.active_validators.items) |val| {
            total_stake += val.stake;
        }

        if (total_stake == 0) return null; // Should not happen

        // We need a bitmask
        const num_vals = self.engine.active_validators.items.len;
        const bitmask_len = (num_vals + 7) / 8;
        var bitmask = try self.allocator.alloc(u8, bitmask_len);
        @memset(bitmask, 0);
        // We defer free? No, we return it.

        // Aggregate BLS
        // var agg_sig: ?*blst.blst_p2 = null; // Unused
        // BLST zig generic wrapper usage:
        // We can use the lower level functions or the high level one in Zelius/VRF?
        // Let's assume we use raw blst_p2_add_or_double logic on Affine->Jacobian

        var acc = std.mem.zeroes(blst.blst_p2);
        var first = true;

        var it = votes_map.iterator();
        while (it.next()) |entry| {
            const idx = entry.key_ptr.*;
            const vote = entry.value_ptr.*;

            if (idx >= num_vals) continue;

            const val = self.engine.active_validators.items[idx];
            voted_stake += val.stake;

            // Set Bitmask
            const byte_idx = idx / 8;
            const bit_idx = @as(u3, @intCast(idx % 8));
            bitmask[byte_idx] |= (@as(u8, 1) << bit_idx);

            // Aggregate Sig
            // deserialize signature
            var sig_affine = std.mem.zeroes(blst.blst_p2_affine);
            const res = blst.blst_p2_uncompress(&sig_affine, &vote.signature);
            if (res != blst.BLST_SUCCESS) continue; // Invalid sig data?

            var sig_jac = std.mem.zeroes(blst.blst_p2);
            blst.blst_p2_from_affine(&sig_jac, &sig_affine);

            if (first) {
                acc = sig_jac;
                first = false;
            } else {
                blst.blst_p2_add_or_double(&acc, &acc, &sig_jac);
            }
        }

        // Check Threshold: > 2/3
        // threshold = (2 * total) / 3
        const threshold = (total_stake * 2) / 3;

        if (voted_stake > threshold) {
            // Serialize AggSig
            var final_affine = std.mem.zeroes(blst.blst_p2_affine);
            blst.blst_p2_to_affine(&final_affine, &acc);

            var sig_bytes: [96]u8 = undefined;
            blst.blst_p2_compress(&sig_bytes, &final_affine);

            return .{ .sig = sig_bytes, .bitmask = bitmask };
        }

        self.allocator.free(bitmask);
        return null;
    }
};
